### PR TITLE
[FLINK-36880][network] Hybrid shuffle supports job master failover if only external tier used

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Hybrid shuffle supports job master failover if only external tier used.*

## Verifying this change

UT & Manually test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredPartitionWithMetrics.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.shuffle;
+
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.ShuffleMetrics;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Partition with shuffle metrics for tiered storage. */
+public class TieredPartitionWithMetrics {
```
Comment (codenohup) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredPartitionWithMetrics.java:27]: The _TieredPartitionWithMetrics_ may be unnecessary, as the _TierShuffleDescriptor_ field is not utilized.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredPartitionWithMetrics.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            TierShuffleDescriptor shuffleDescriptor, ShuffleMetrics partitionMetrics) {
+        this.shuffleDescriptor = checkNotNull(shuffleDescriptor);
+        this.partitionMetrics = checkNotNull(partitionMetrics);
+    }
+
+    public ShuffleMetrics getPartitionMetrics() {
+        return partitionMetrics;
+    }
+
+    public TierShuffleDescriptor getPartition() {
```
Comment (codenohup) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredPartitionWithMetrics.java:41]: This is a weird name for a method, why _getPartition_ return _TierShuffleDescriptor_

**Review by codenohup — COMMENTED** [review]
Thank you for your contribution, I have a few minor questions, PTAL.

**Review by reswqa — COMMENTED** [review]
Thanks for the review, updated.

**Review by codenohup — APPROVED** [review]
LGTM, thanks for the update.

**Pull request merged**